pkg/recipe/crd.go

```diff
@@ -0,0 +1,248 @@
+	return strings.HasSuffix(apiVersion, "/"+version)
+}
+
+func (e *ExecutableCRD) postCreate() error {
+	message := fmt.Sprintf(
+		"PostCreate CRD: Resource %s: APIVersion %s",
+		e.CRResource,
+		e.CRAPIVersion,
+	)
+	// discover custom resource API
+	err := e.Retry.Waitf(
+		func() (bool, error) {
+			got := e.GetAPIForAPIVersionAndResource(
+				e.CRAPIVersion,
+				e.CRResource,
+			)
+			if got == nil {
+				return e.IsFailFastOnDiscoveryError(),
+					errors.Errorf(
+						"Failed to discover CRD: Resource %s: APIVersion %s",
+						e.CRResource,
+						e.CRAPIVersion,
+					)
+			}
+			// fetch dynamic client for the custom resource
+			// corresponding to this CRD
+			cli, err := e.GetClientForAPIVersionAndResource(
+				e.CRAPIVersion,
+				e.CRResource,
+			)
+			if err != nil {
+				return e.IsFailFastOnDiscoveryError(), err
+			}
+			// A successful list implies CRD is registered & is discovered
+			// in this binary
+			_, err = cli.List(metav1.ListOptions{})
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		},
+		message,
+	)
+	return err
+}
+
+// Create creates the CRD in Kubernetes cluster
+func (e *ExecutableCRD) Create() (*types.CreateResult, error) {
+	cli, err := e.dynamicClientset.GetClientForAPIVersionAndKind(
+		e.State.GetAPIVersion(), // CRD APIVersion & not CR APIVersion
+		e.State.GetKind(),       // CRD Kind & not CR Kind
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = cli.Create(e.State, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// add to teardown functions
+	e.AddToTeardown(func() error {
+		_, err := cli.Get(
+			e.State.GetName(),
+			metav1.GetOptions{},
+		)
+		if err != nil && apierrors.IsNotFound(err) {
+			// nothing to do
+			return nil
+		}
+		return cli.Delete(
+			e.State.GetName(),
+			nil,
+		)
+	})
+	if !e.IgnoreDiscovery {
+		// run an additional step to wait till this CRD
+		// is discovered at apiserver
+		err = e.postCreate()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &types.CreateResult{
+		Phase: types.CreateStatusPassed,
+		Message: fmt.Sprintf(
+			"Create CRD: Resource %s: APIVersion %s",
+			e.CRResource,
+			e.CRAPIVersion,
+		),
+	}, nil
+}
+
+// Update performs a 3-way merge of the CRD in the Kubernetes cluster
+func (e *ExecutableCRD) Update() (*types.ApplyResult, error) {
+	cli, err := e.dynamicClientset.GetClientForAPIVersionAndKind(
+		e.State.GetAPIVersion(), // CRD APIVersion & not CR APIVersion
+		e.State.GetKind(),       // CRD Kind & not CR Kind
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the CRD observed at the cluster
+	target, err := cli.Get(
+		e.State.GetName(),
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	merged := &unstructured.Unstructured{}
+	// 3-way merge
+	merged.Object, err = dynamicapply.Merge(
+		target.UnstructuredContent(),  // observed
+		e.State.UnstructuredContent(), // last applied
+		e.State.UnstructuredContent(), // desired
```

Comment:
3-way merge is possible on managed fields only. _This is how multiple  applications (read controllers can update the same resource without conflicts)._ However, d-operators does not manage any fields like kubectl or metacontroller _(refer to last applied state annotations)_. 

d-operators needs to merge the desired state without having to know what is the current observed state. Since, 3-way merge is the best way to achieve this. d-operators makes a false claim that last applied was managed by itself by setting desired state to last applied. 

This also mean that clients/code using d-operators should only set the precise fields they are interested in.